### PR TITLE
Block - perform some validations when parsing the request

### PR DESCRIPTION
**Pull request by serban300**
# Reason for This PR

Block - perform some validations when parsing the request

Addresses a comment from an older PR: https://github.com/firecracker-microvm/firecracker/pull/2629#discussion_r656187051

## Description of Changes

This simplifies the request execution logic which will make it easier for us to add io_uring support.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: src/devices/src/virtio/block/request.rs [hunk 1]
```diff
@@ -656,22 +656,26 @@ mod tests {
...
         (
-            any::<u64>(), // random data buffer sparsity factor
-            any::<u32>(), // data_len
-            any::<u64>(), // sector
-            any::<RequestType>(),
+            any::<u64>(),         // random data buffer sparsity factor
+            any::<u32>(),         // data_len
+            any::<u64>(),         // sector
+            any::<RequestType>(), // request type
+            any::<[bool; 32]>(),  // coin
```
Comment (alindima) [inline src/devices/src/virtio/block/request.rs]: why 32 coins? AFAICT 5 coins are used

File: src/devices/src/virtio/block/request.rs [hunk 2]
```diff
@@ -424,152 +424,213 @@ mod tests {
         }
     }
 
-    #[test]
-    #[allow(clippy::cognitive_complexity)]
-    fn test_parse() {
-        let m = &GuestMemoryMmap::from_ranges(&[(GuestAddress(0), 0x10000)]).unwrap();
-        let vq = VirtQueue::new(GuestAddress(0), &m, 16);
+    struct RequestVirtQueue<'a> {
```
Comment (alindima) [inline src/devices/src/virtio/block/request.rs:427]: Nice abstraction! The naming is not expressive enough, though I couldn't find a better one.

File: src/devices/src/virtio/block/request.rs [hunk 3]
```diff
@@ -840,24 +883,58 @@ mod tests {
...
+                // Writeable buffer is readonly.
+                RequestType::In => {
+                    data_desc_flags.set(data_desc_flags.get() & !VIRTQ_DESC_F_WRITE);
+                    return (Err(Error::UnexpectedReadOnlyDescriptor), mem, q);
+                }
+                // Writeable buffer is readonly.
+                RequestType::GetDeviceID => {
+                    data_desc_flags.set(data_desc_flags.get() & !VIRTQ_DESC_F_WRITE);
+                    return (Err(Error::UnexpectedReadOnlyDescriptor), mem, q);
+                }
```
Comment (alindima) [inline src/devices/src/virtio/block/request.rs:902]: These two branches are identical

Comment (serban300) [reply src/devices/src/virtio/block/request.rs:902]: Nice catch. Fixed.

File: src/devices/src/virtio/block/request.rs (hunk 2 above)
Comment (serban300) [reply src/devices/src/virtio/block/request.rs:427]: I can't think of a better name now. Anyway we can change it in the future.

File: src/devices/src/virtio/block/request.rs (hunk 1 above)
Comment (serban300) [reply src/devices/src/virtio/block/request.rs]: True. I would like to have a margin of error anyway. I reduced the number of coins to 10.

**Review by alindima — APPROVED** [review]

**Review by acatangiu — APPROVED** [review]
Looks good!

**Pull request merged**